src/dataflow/src/render/join/linear_join.rs

```diff
@@ -77,35 +213,51 @@ where
                 .filter(filter)
                 .project(project);
 
-            // Construct initial join build state.
-            // This state will evolves as we build the join dataflow.
-            let mut join_build_state = JoinBuildState::new(
-                input_mapper.global_columns(*start),
-                &equivalences,
-                &map_filter_project,
+            // Create a join plan based on the linear join.
+            let linear_plan = LinearJoinPlan::create_from(
+                *start,
+                equivalences,
+                order,
+                input_mapper,
+                map_filter_project,
             );
 
-            // This collection will evolve as we join in more inputs.
-            // TODO(mcsherry): determine and apply closure here in `flat_map_ref` form.
-            // TODO(mcsherry): If we plan to use an arrangement, should one exist, then
-            // this is wasteful as it instantiates all rows which are then dropped.
-            let (mut joined, mut errs) = self.collection(&inputs[*start]).unwrap();
-
-            let use_leading_arrangement = start_arr.is_some() && inputs.len() > 1;
-            if !use_leading_arrangement {
-                // NOTE(mcsherry): ideally this code is rarely/never relevant, as the associated logic
-                // could be pushed down to the input and perhaps beyond. I'm not certain under what
-                // circumstance we should just delete it, though.
-
-                // At this point we are able to construct a per-row closure that can be applied once
-                // we have the first wave of columns in place. We will not apply it quite yet, because
-                // we have three code paths that might produce data and it is complicated.
-                let closure = join_build_state.extract_closure();
-                if !closure.is_identity() {
+            // Collect all error streams, and concatenate them at the end.
+            let mut errors = Vec::new();
+
+            // Determine what form the linear spine of updates will take.
+            // This could be a stream, or one of several arrangements.
+            let mut joined = if start_arr.is_some() && inputs.len() > 1 {
+                // Assert that we did not extract an initial closure.
+                assert!(linear_plan.initial_closure.is_none());
+                // We have some redundancy that we should check here.
+                let arrangement_key = &linear_plan.stage_plans[0].stream_key;
+                assert_eq!(start_arr.as_ref().unwrap(), arrangement_key);
```

Comment:
This assertion is hit:

```
materialized v0.7.1-dev (7c7b794b8) listening on 0.0.0.0:6875...
2021-03-03T10:33:40.413887Z ERROR panic: assertion failed: `(left == right)`
  left: `[]`,
 right: `[Column(0)]`
thread: timely:work-8
location: src/dataflow/src/render/join/linear_join.rs:235:17
version: 0.7.1-dev (7c7b794b8fa3af3015fe939a9a4596cd18638110)
```


Test case:

```
CREATE TABLE orders (
    o_orderkey       integer,
    o_custkey        integer NOT NULL,
    o_orderdate      date NOT NULL
);

CREATE INDEX pk_orders_orderkey ON orders (o_orderkey);
CREATE INDEX fk_orders_custkey ON orders (o_custkey ASC);

CREATE TABLE lineitem (

    l_commitdate     date NOT NULL,
    l_receiptdate    date NOT NULL
);

SELECT * FROM lineitem  , orders
WHERE o_custkey  IN (
	SELECT  o_custkey FROM lineitem RIGHT JOIN orders ON ( true )
	WHERE o_orderkey < 1460
	AND l_receiptDATE = '1996-11-23'
	OR l_commitDATE = o_orderdate
) ;```

Reply:
Thanks!

Reply:
In looking in to this, I *think* there is a disparity between the `explain plan` and the rendered join. At least, the explain plan for this query has no joins (that I can see) which have `start_arr.is_some()`. The explain plan is:
```
-----------------------------------------------------------------------
 %0 =                                                                 +
 | Get materialize.public.lineitem (u5)                               +
 | ArrangeBy ()                                                       +
                                                                      +
 %1 =                                                                 +
 | Get materialize.public.orders (u1)                                 +
                                                                      +
 %2 = Let l0 =                                                        +
 | Join %0 %1                                                         +
 | | implementation = Differential %1 %0.()                           +
 | | demand = (#0..#4)                                                +
                                                                      +
 %3 =                                                                 +
 | Get %2 (l0)                                                        +
 | Distinct group=(#3)                                                +
 | ArrangeBy ()                                                       +
                                                                      +
 %4 =                                                                 +
 | Get materialize.public.orders (u1)                                 +
                                                                      +
 %5 = Let l1 =                                                        +
 | Join %3 %4                                                         +
 | | implementation = Differential %4 %3.()                           +
 | | demand = (#0..#3)                                                +
                                                                      +
 %6 =                                                                 +
 | Get materialize.public.lineitem (u5)                               +
 | ArrangeBy ()                                                       +
                                                                      +
 %7 = Let l2 =                                                        +
 | Join %6 %5                                                         +
 | | implementation = Differential %5 %6.()                           +
 | | demand = (#0..#5)                                                +
                                                                      +
 %8 =                                                                 +
 | Get %7 (l2)                                                        +
 | Filter (((#3 < 1460) && (#1 = 1996-11-23)) || (#0 = #5)), (#2 = #4)+
 | Project (#2, #0, #1, #3..#5)                                       +
                                                                      +
 %9 =                                                                 +
 | Get %7 (l2)                                                        +
 | Distinct group=(#2, #3, #4, #5)                                    +
 | Negate                                                             +
 | Filter (((#1 < 1460) && null) || null), (#0 = #2)                  +
                                                                      +
 %10 =                                                                +
 | Get %5 (l1)                                                        +
 | Filter (((#1 < 1460) && null) || null), (#0 = #2)                  +
 | Distinct group=(#0, #1, #2, #3)                                    +
                                                                      +
 %11 =                                                                +
 | Union %9 %10                                                       +
 | Filter (((#1 < 1460) && null) || null), (#0 = #2)                  +
 | ArrangeBy (#0, #1, #2, #3)                                         +
                                                                      +
 %12 =                                                                +
 | Get %5 (l1)                                                        +
 | Filter (((#1 < 1460) && null) || null), (#0 = #2)                  +
                                                                      +
 %13 =                                                                +
 | Join %11 %12 (= #0 #4) (= #1 #5) (= #2 #6) (= #3 #7)               +
 | | implementation = Differential %12 %11.(#0, #1, #2, #3)           +
 | | demand = (#0..#3)                                                +
 | Map null, null                                                     +
 | Project (#0, #8, #9, #1..#3)                                       +
                                                                      +
 %14 =                                                                +
 | Union %8 %13                                                       +
 | Filter (((#3 < 1460) && (#2 = 1996-11-23)) || (#1 = #5)), (#0 = #4)+
 | Distinct group=(#0)                                                +
 | ArrangeBy (#0)                                                     +
                                                                      +
 %15 =                                                                +
 | Join %2 %14 (= #3 #5)                                              +
 | | implementation = Differential %2 %14.(#0)                        +
 | | demand = (#0..#4)                                                +
 | Project (#0..#4)                                                   +
 
(1 row)
```
So, it does seem like there is a mismatch between `start_arr` that gets introduced (further optimization, in the dataflow builder perhaps?), although the prior code just ignored it and used the equivalent of `linear_plan.stage_plans[0].stream_key` (looking at the keys of the opposing relation and determining a necessary key for the first input).

In particular, `start_arr` and `join_order` are for this join (respectively):
```
Some([]) 
[(2, [Column(1)]), (0, [])]
```
which seems wrong, as that first join should satisfy the invariant that `start_arr.len()` is the same as the first length. @wangandi I'm inclined to just comment out the use of `start_arr` though we should probably look in to whether it is being properly determined (as it wasn't used in rendering, it could have been incorrectly set or maintained for a while).

Reply:
This assertion is no longer seen.